### PR TITLE
Tag feature

### PR DESCRIPTION
**Pull request by 1mn**
image

### CODE REVIEW

**Comment by hargoniX** [comment]
Soo its seems snekchek will try to run all linters if there is no config file existent, that is what causes CI to fail I guess, let me try to fix (Mart will not publish to pypi today anymore I guess)
Also why do you still have snekchek in CI if you deleted .snekrc?

**Comment by martmists** [comment]
Why did you remove .snekrc if we're using snekchek for flake8 and soon pytest? If you want Travis to pass, simply set .travis.yml to use flake8 instead, but I'm confused as to why.

File: bot/cogs/clickup.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def __init__(self, bot: AutoShardedBot):
         self.lists = CaseInsensitiveDict()
 
     async def on_ready(self):
-        with ClientSession() as session:
+        async with ClientSession() as session:
```
Comment (1mn) [inline bot/cogs/clickup.py]: These are out of scope. Should I remove them?

Comment (1mn) [reply bot/cogs/clickup.py]: they are probably improvements, but maybe it'd be better if gdude made the changes to this file.

Comment (1mn) [reply bot/cogs/clickup.py]: For the record, I did these so I could get aiohttp3 to work, but then later realized I should be using 2.5 with this version of discord.py rewrite.

Comment (1mn) [reply bot/cogs/clickup.py]: eh, I'll remove them.

File: bot/cogs/tags.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+
+class Tags:
+    """
+    Save new tags and fetch existing tags.
+    """
+
+    def __init__(self, bot: AutoShardedBot):
+        self.bot = bot
+        self.api_key = os.environ.get("BOT_API_KEY")
```
Comment (gdude2002) [inline bot/cogs/tags.py]: `constants.SITE_API_KEY`

File: bot/cogs/tags.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        self.tag_cooldowns = {}
+
+    async def get_tag_data(self, tag_name: Union[str, None] = None) -> dict:
```
Comment (gdude2002) [inline bot/cogs/tags.py]: You can replace `Union[anything, None]` with `Optional[anything]`

File: bot/cogs/tags.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    async def info(self, ctx: Context):
+        """
+        Show available methods for this class.
+
+        :param ctx: Discord message context
+        """
+
+        return await ctx.invoke(self.bot.get_command("help"), "Tags")
+
+    @command(name="tags.get()", aliases=["tags.get", "tags.show()", "tags.show"])
```
Comment (gdude2002) [inline bot/cogs/tags.py]: So far, I've been adding command aliases that make sense without the namespace, but that's entirely up to you I think. Like, ClickUp has `open_task` for example.

File: bot/cogs/tags.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    async def get(self, ctx: Context, tag_name: str = None):
+        """
+        Get a list of all tags or a specified tag.
+
+        :param ctx: Discord message context
```
Comment (gdude2002) [inline bot/cogs/tags.py]: Is this a thing we're doing in command docstrings? It's a good idea, I think

File: bot/cogs/tags.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        :param tag_name: The name of the tag to create or edit.
+        :param tag_content: The content of the tag.
+        """
+
+        embed = Embed()
+
+        if "\n" in tag_name:
+            embed.colour = Colour.red()
+            embed.title = "Please don't do that"
+            embed.description = "Don't be ridiculous. Newlines are obviously not allowed in the tag name."
```
Comment (gdude2002) [inline bot/cogs/tags.py]: 👍

File: bot/cogs/tags.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+        else:
+            if tag_name and tag_content:
```
Comment (gdude2002) [inline bot/cogs/tags.py]: This could be reorganised, for example since the `else` here contains a return, you could invert the conditional and drop the extra branch.

```py
if not (tag_name and tag_content):
    # ....
    return

tag_data = ...
```

File: bot/cogs/tags.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                embed.title = "Missing parameters!",
+                embed.description = "The tag needs both a name and some content"
+                return await ctx.send(embed=embed)
+
+            if tag_data.get("success"):
+                embed.colour = Colour.blurple()
+                embed.title = "Tag successfully added!"
+                embed.description = f"**{tag_name}** added to tag database."
+            else:
+                print(f"Something terrible happened. The API returned {tag_data}")
```
Comment (gdude2002) [inline bot/cogs/tags.py]: It's bad UX to just return the json - it should be clear what an error state will give us, and we should be able to provide a useful error message

File: bot/cogs/tags.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        """
+        Remove a tag from the database.
+
+        :param ctx: discord message context
+        :param tag_name: The name of the tag to delete.
+        """
+
+        embed = Embed()
+
+        if tag_name:
```
Comment (gdude2002) [inline bot/cogs/tags.py]: Again, this conditional could be inverted in order to drop a branch

File: bot/cogs/tags.py (hunk 5 above)
Comment (1mn) [reply bot/cogs/tags.py]: I don't see why not, but I think it was mostly just something pycharm did :P

Comment (gdude2002) [reply bot/cogs/tags.py]: It's basically Sphinx markup - the sphinx document generator uses it when it's generating documentation from your code.

I think it is a good idea to explain the params to the users as well, though.

File: bot/cogs/tags.py (hunk 8 above)
Comment (1mn) [reply bot/cogs/tags.py]: yeah this was more for debugging. forgot I left it in.

File: bot/cogs/tags.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            embed.title = "Please don't do that"
+            embed.description = "Don't be ridiculous. Newlines are obviously not allowed in the tag name."
+
+        else:
+            if tag_name and tag_content:
+                tag_name = tag_name.lower()
+                tag_data = await self.post_tag_data(tag_name, tag_content)
+
+            else:
+                embed.colour = Colour.red(),
```
Comment (1mn) [inline bot/cogs/tags.py]: lol wtf is this comma

File: bot/__main__.py [hunk 11]
```diff
@@ -34,5 +34,6 @@
 bot.load_extension("bot.cogs.fun")
 bot.load_extension("bot.cogs.eval")
 bot.load_extension("bot.cogs.verification")
+bot.load_extension("bot.cogs.tags")
```
Comment (AnonGuy) [inline bot/__main__.py]: Can these be in alphabetical size order pls

File: bot/cogs/tags.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        Show available methods for this class.
+
+        :param ctx: Discord message context
+        """
+
+        return await ctx.invoke(self.bot.get_command("help"), "Tags")
+
+    @command(name="tags.get()", aliases=["tags.get", "tags.show()", "tags.show", "get_tag"])
+    async def get_command(self, ctx: Context, tag_name: str = None):
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: Why isn't `tag_name` type hinted as an `Optional[str]`?

File: bot/cogs/tags.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        Get a list of all tags or a specified tag.
+
+        :param ctx: Discord message context
+        :param tag_name:
+        If provided, this function shows data for that specific tag.
+        If not provided, this function shows the caller a list of all tags.
+        """
+
+        def _command_on_cooldown(tag_name: Optional[str]) -> bool:
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: Why is `tag_name` type hinted as an `Optional`? It doesn't have a default value.

File: bot/cogs/tags.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+            """
+
+            now = time.time()
+
+            cooldown_conditions = (
+                tag_name
+                and tag_name in self.tag_cooldowns
+                and (now - self.tag_cooldowns[tag_name]["time"]) < TAG_COOLDOWN
+                and self.tag_cooldowns[tag_name]["channel"] == ctx.channel.id
+            )
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: This is :ok_hand:. Will write my conditions like this from now on.

File: bot/cogs/tags.py [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+class Tags:
+    """
+    Save new tags and fetch existing tags.
+    """
+
+    def __init__(self, bot: AutoShardedBot):
+        self.bot = bot
+        self.tag_cooldowns = {}
+
+    async def get_tag_data(self, tag_name: Optional[str] = None) -> dict:
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: should there be spaces around the `=` in this keyword assignment?

File: bot/cogs/tags.py (hunk 12 above)
Comment (AnonGuy) [reply bot/cogs/tags.py]: Also shouldn't keyword assignments have no spaces around the `=`?

File: bot/cogs/tags.py [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        Retrieve the tag_data from our API
+
+        :param tag_name: the tag to retrieve
+        :return:
+        if tag_name was provided, returns a dict with tag data.
+        if not, returns a list of dicts with all tag data.
+
+        """
+        headers = {"X-API-KEY": SITE_API_KEY}
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: You're assigning this headers dict with the same value in 3 different functions.
Maybe move it outside of the function scope?

File: bot/__main__.py [hunk 17]
```diff
@@ -35,5 +35,6 @@
 bot.load_extension("bot.cogs.eval")
 bot.load_extension("bot.cogs.fun")
 bot.load_extension("bot.cogs.verification")
+bot.load_extension("bot.cogs.tags")
```
Comment (AnonGuy) [inline bot/__main__.py]: Can these be in size order pl0x

File: bot/cogs/tags.py [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+
+            tag_name = tag_name.lower()
+            tag_data = await self.post_tag_data(tag_name, tag_content)
+
+            if tag_data.get("success"):
+                embed.colour = Colour.blurple()
+                embed.title = "Tag successfully added!"
+                embed.description = f"**{tag_name}** added to tag database."
+            else:
+                embed.title = "Database error",
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: Some of the embed titles have punctuation but some don't. It might be more consistent to add periods to these?

File: bot/cogs/tags.py [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+        embed.colour = Colour.red()
+
+        if "\n" in tag_name:
+            embed.title = "Please don't do that"
+            embed.description = "Don't be ridiculous. Newlines are obviously not allowed in the tag name."
+
+        else:
+            if not (tag_name and tag_content):
+                embed.title = "Missing parameters!"
+                embed.description = "The tag needs both a name and some content"
```
Comment (AnonGuy) [inline bot/cogs/tags.py]: Can we have a period at the end of this description pl0x
